packages/web3-common/src/formatter.ts

```diff
@@ -0,0 +1,254 @@
+}
+
+export type ByteTypes = {
+	[FMT_BYTES.HEX]: HexString;
+	[FMT_BYTES.BUFFER]: Buffer;
+	[FMT_BYTES.UINT8ARRAY]: Uint8Array;
+};
+
+export type DataFormat = {
+	readonly number: FMT_NUMBER;
+	readonly bytes: FMT_BYTES;
+};
+
+export const DEFAULT_RETURN_FORMAT = { number: FMT_NUMBER.HEX, bytes: FMT_BYTES.HEX } as const;
+
+// Added `undefined` to cover optional type
+export type FormatType<T, F extends DataFormat> = number extends Extract<T, Numbers>
+	? NumberTypes[F['number']] | Exclude<T, Numbers>
+	: Buffer extends Extract<T, Bytes>
+	? ByteTypes[F['bytes']] | Exclude<T, Bytes>
+	: T extends object | undefined
+	? {
+			[P in keyof T]: FormatType<T[P], F>;
+	  }
+	: T;
+
+const findSchemaByDataPath = (schema: JsonSchema, dataPath: string[]): JsonSchema | undefined => {
+	let result: JsonSchema = { ...schema } as JsonSchema;
+
+	for (const dataPart of dataPath) {
+		if (!result.properties && !result.items) {
+			return undefined;
+		}
+
+		if (result.properties) {
+			result = (result.properties as Record<string, JsonSchema>)[dataPart];
+		} else if (result.items && (result.items as JsonSchema).properties) {
+			const node = (result.items as JsonSchema).properties as Record<string, JsonSchema>;
+
+			if (!node) {
+				return undefined;
+			}
+
+			result = node[dataPart];
+		} else if (result.items && isObject(result.items)) {
+			result = result.items;
+		} else if (result.items && Array.isArray(result.items)) {
+			result = (result.items as JsonSchema[])[parseInt(dataPart, 10)];
+		}
+	}
+
+	return result;
+};
+
+export const convertScalarValue = (value: unknown, ethType: string, format: DataFormat) => {
+	try {
+		const { baseType } = parseBaseType(ethType);
+
+		if (baseType === 'int' || baseType === 'uint') {
+			switch (format.number) {
+				case FMT_NUMBER.NUMBER:
+					return Number(toBigInt(value));
+				case FMT_NUMBER.HEX:
+					return numberToHex(toBigInt(value));
+				case FMT_NUMBER.STR:
+					return toBigInt(value).toString();
+				case FMT_NUMBER.BIGINT:
+					return toBigInt(value);
+				default:
+					throw new FormatterError(`Invalid format: ${String(format.number)}`);
+			}
+		}
+
+		if (baseType === 'bytes') {
+			switch (format.bytes) {
+				case FMT_BYTES.HEX:
+					return bytesToHex(bytesToBuffer(value as Bytes));
+				case FMT_BYTES.BUFFER:
+					return bytesToBuffer(value as Bytes);
+				case FMT_BYTES.UINT8ARRAY:
+					return new Uint8Array(bytesToBuffer(value as Bytes));
+				default:
+					throw new FormatterError(`Invalid format: ${String(format.bytes)}`);
+			}
+		}
+	} catch (error) {
+		// TODO: Add debugging to verify the error is thrown by the correct function
+		return value;
+	}
+
+	return value;
+};
+
+export const convert = (
+	data: Record<string, unknown> | unknown[] | unknown,
+	schema: JsonSchema,
+	dataPath: string[],
+	format: DataFormat,
+) => {
+	// If it's a scalar value
+	if (!isObject(data) && !Array.isArray(data)) {
+		return convertScalarValue(data, schema?.eth as string, format);
```

Comment:
Since it's possible for `schema.eth` to be undefined, would this cause an error to be thrown in `convertScalarValue`?